k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java

```diff
@@ -652,72 +709,79 @@ public void fetch(List<ImapMessage> messages, FetchProfile fetchProfile,
                         .build();
-                command.send(connection);
 
-                ImapResponse response;
-                int messageNumber = 0;
+                sendAndProcessUidFetchCommand(command, messageMap, listener);
+            } catch (IOException ioe) {
+                throw ioExceptionHandler(connection, ioe);
+            }
+        }
+    }
 
-                do {
+    private void sendAndProcessUidFetchCommand(UidFetchCommand command, HashMap<String, Message> messageMap,
+            MessageRetrievalListener<ImapMessage> listener) throws MessagingException, IOException {
+        command.send(connection);
 
-                    response = command.readResponse(connection);
-
-                    if (response.getTag() == null && ImapResponseParser.equalsIgnoreCase(response.get(1), "FETCH")) {
-                        ImapList fetchList = (ImapList) response.getKeyedValue("FETCH");
-                        String uid = fetchList.getKeyedString("UID");
-                        long msgSeq = response.getLong(0);
-                        if (uid != null) {
-                            try {
-                                msgSeqUidMap.put(msgSeq, uid);
-                                if (K9MailLib.isDebug()) {
-                                    Timber.v("Stored uid '%s' for msgSeq %d into map", uid, msgSeq);
-                                }
-                            } catch (Exception e) {
-                                Timber.e("Unable to store uid '%s' for msgSeq %d", uid, msgSeq);
-                            }
-                        }
+        ImapResponse response;
+        int messageNumber = 0;
 
-                        Message message = messageMap.get(uid);
-                        if (message == null) {
-                            if (K9MailLib.isDebug()) {
-                                Timber.d("Do not have message in messageMap for UID %s for %s", uid, getLogId());
-                            }
+        do {
 
-                            handleUntaggedResponse(response);
-                            continue;
-                        }
+            response = command.readResponse(connection);
 
-                        if (listener != null) {
-                            listener.messageStarted(uid, messageNumber++, messageMap.size());
+            if (response.getTag() == null && ImapResponseParser.equalsIgnoreCase(response.get(1), "FETCH")) {
+                ImapList fetchList = (ImapList) response.getKeyedValue("FETCH");
+                String uid = fetchList.getKeyedString("UID");
+                long msgSeq = response.getLong(0);
+                if (uid != null) {
+                    try {
+                        msgSeqUidMap.put(msgSeq, uid);
+                        if (K9MailLib.isDebug()) {
+                            Timber.v("Stored uid '%s' for msgSeq %d into map", uid, msgSeq);
                         }
+                    } catch (Exception e) {
```

Comment:
Funny, the `ConcurrentHashMap.put()` method does not throw any checked exceptions, so this try/catch block is useless. I'm not sure why this was there in the first place. I moved around a lot of original code without modifying it.